bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOIntegrationTest.java

```diff
@@ -102,12 +111,108 @@ public void testWriteToTable_tableExist() throws Exception {
   }
 
   @Test
-  public void testEstimatedAndSplitForSmallTable() throws Exception {
+  public <T> void testWriteToTable_dataWritten() throws Exception {
+    int INSERT_COUNT = 50;
+    try (Admin admin = connection.getAdmin()) {
+      LOG.info("Creating table in testWriteToTable_dataWritten()");
+      TableName tableName = createNewTable(admin);
+      try {
+        writeThroughDataflow(tableName, INSERT_COUNT);
+        checkTableRowCount(tableName, INSERT_COUNT);
+      } finally {
+        admin.deleteTable(tableName);
+      }
+    }
+  }
+
+  protected TableName createNewTable(Admin admin) throws IOException {
     TableName tableName = newTestTableName();
+    admin.createTable(
+      new HTableDescriptor(tableName).addFamily(new HColumnDescriptor(COLUMN_FAMILY)));
+    return tableName;
+  }
+
+  protected void writeThroughDataflow(TableName tableName, int insertCount) throws Exception {
+    CloudBigtableIO.CloudBigtableSingleTableWriteFn writer =
+        new CloudBigtableIO.CloudBigtableSingleTableWriteFn(
+            createConfig(tableName.getNameAsString()));
+    @SuppressWarnings("unchecked")
+    DoFn<Mutation, Void>.ProcessContext mockContext = mock(ProcessContext.class);
+    final AtomicInteger counter = new AtomicInteger();
+    when(mockContext.element()).thenAnswer(new Answer<Put>() {
+      @Override
+      public Put answer(InvocationOnMock invocation) throws Throwable {
+        byte[] row = Bytes.toBytes("row_" + counter.incrementAndGet());
+        return new Put(row).addColumn(COLUMN_FAMILY, QUALIFIER1,
+          Bytes.toBytes(RandomStringUtils.randomAlphanumeric(8)));
+      }
+    });
+    for (int i = 0; i < insertCount; i++) {
+      writer.processElement(mockContext);
+    }
+    writer.finishBundle(null);
+  }
+
+  @SuppressWarnings("unused")
+  protected void checkTableRowCount(TableName tableName, int rowCount) throws IOException {
+    int readCount = 0;
+    try (Table table = connection.getTable(tableName);
+        ResultScanner scanner = table.getScanner(new Scan())) {
+      for (Result result : scanner) {
+        readCount++;
+      }
+    }
+    Assert.assertEquals(rowCount, readCount);
+  }
+
+  @Test
+  public <T> void testReadFromTable_singleResultDataflowReader() throws Exception {
+    int INSERT_COUNT = 50;
+    try (Admin admin = connection.getAdmin()) {
+      TableName tableName = createNewTable(admin);
+      try {
+        writeViaTable(tableName, INSERT_COUNT);
+        checkTableRowCountViaDataflowResultReader(tableName, INSERT_COUNT);
+      } finally {
+        admin.deleteTable(tableName);
+      }
+    }
+  }
+
+  protected void writeViaTable(TableName tableName, int rowCount) throws IOException {
```

Comment:
Why writeViaTable is protected but checkTableRowCountViaDataflowResultReader is private?
Can we make them all private if they are all indented to be used in this test class?